core-bundle/src/EventListener/DataContainer/Undo/LabelListener.php

```diff
@@ -0,0 +1,229 @@
+            Date::parse(Config::get('dateFormat'), $row['tstamp']),
+            Date::parse(Config::get('timeFormat'), $row['tstamp']),
+        );
+
+        $header .= sprintf(
+            '<div class="tl_undo_header_item"><span class="tl_undo_header_label">%s</span> <strong>%s</strong></div>',
+            $GLOBALS['TL_LANG']['tl_undo']['pid'][0],
+            $user ? $user->username : $row['pid'],
+        );
+
+        $header .= sprintf(
+            '<div class="tl_undo_header_item"><span class="tl_undo_header_label">%s</span> <strong>%s</strong></div>',
+            $GLOBALS['TL_LANG']['tl_undo']['fromTable'][0],
+            $type,
+        );
+
+        $header .= sprintf(
+            '<div class="tl_undo_header_item"><span class="tl_undo_header_label">ID</span> %s</div>',
+            $originalRow['id'],
+        );
+
+        if ($parent) {
+            $header .= sprintf(
+                '<div class="tl_undo_header_item"><span class="tl_undo_header_label">%s</span> <strong>%s</strong></div>',
+                $GLOBALS['TL_LANG']['tl_undo']['parent'],
+                $this->getTranslatedTypeFromTable($parent['table'])
+            );
+        }
+
+        $header .= '</div>';
+
+        return $header;
+    }
+
+    /**
+     * @throws Exception
+     * @throws DriverException
+     *
+     * @return array|string
+     */
+    private function renderLabel(array $arrRow, DataContainer $dc)
+    {
+        $mode = $GLOBALS['TL_DCA'][$dc->table]['list']['sorting']['mode'] ?? DataContainer::MODE_SORTED;
+
+        if (DataContainer::MODE_PARENT === $mode) {
+            $callback = $GLOBALS['TL_DCA'][$dc->table]['list']['sorting']['child_record_callback'] ?? null;
+
+            if (\is_array($callback)) {
+                $callable = System::importStatic($callback[0]);
+
+                return $callable->{$callback[1]}($arrRow);
+            }
+
+            if (\is_callable($callback)) {
+                return $callback($arrRow);
+            }
+        }
+
+        $labelConfig = &$GLOBALS['TL_DCA'][$dc->table]['list']['label'];
+        $labelValues = [];
+
+        foreach ($labelConfig['fields'] as $k => $v) {
+            if (false !== strpos($v, ':')) {
+                [$strKey, $strTable] = explode(':', $v);
+                [$strTable, $strField] = explode('.', $strTable);
+
+                $objRef = $this->connection
+                    ->prepare('SELECT '.$strField.' FROM '.$strTable.' WHERE id=? LIMIT 1')
+                    ->executeQuery($arrRow[$strKey])
+                ;
+
+                $labelValues[$k] = $objRef->rowCount() ? $objRef->{$strField} : '';
+            } elseif (\in_array($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['flag'], [DataContainer::SORT_DAY_ASC, DataContainer::SORT_DAY_DESC, DataContainer::SORT_MONTH_ASC, DataContainer::SORT_MONTH_DESC, DataContainer::SORT_YEAR_ASC, DataContainer::SORT_YEAR_DESC], true)) {
```

Comment:
Please use early returns in favor of `elseif`/`else` clauses for better readability.

Reply:
I agree this is ugly legacy style in a listener (I copied it from the Picker widget class), but before I start refactoring this, I wait for #3644 and its `generateRecordLabel ` method to settle, because that already would handle some of this. 

Also I wonder if [Haste\ModelData](https://github.com/codefog/contao-haste/blob/main/library/Haste/Generator/ModelData.php#L35-L90) is not something we might add to the Core bundle? Right now, every feature/implementation has to carefully handle a DCA field on its own – decoding, encoding, formatting… However this would be something for another PR ;-)